### DISCUSSION TITLE
FlatConfig: Ignore everything except a few files

### DISCUSSION THREAD

**Discussion by canassa** [eslint/eslint#18667] [Q&A] [answered]
I am using v9 and the FlatConfig, but I am unable to make ESLint ignore everything but a few files. This is what I am trying to do:

```js
    {
        ignores: [
            '**/*',
            '!src/functions/health/get.js'
        ],
    },
```
    
However, the get.js isn't unignored and I get this error:
    
```
Oops! Something went wrong! :(

ESLint: 9.6.0

You are linting ".", but all of the files matching the glob pattern "." are ignored.
```

However, it's able to unignore files that are inside in the root dir:

```js
    {
        ignores: [
            '**/*',
            '!src/functions/health/get.js',
            '!prettier.config.js'   // works
        ],
    },
```

EDIT: 

I found a workaround, but it's not ideal:

```
{
    ignores: [
        '*',
        '!src',
        'src/**/*.js'
    ],
}
```

**Comment by mdjermanovic** [comment]
It's the same as with .gitignore: patterns can match directories, and files in ignored directories are considered ignored. So, `**/*` also matches directories `src/`, `src/functions/`, and `src/functions/health/`, and thus you'll need to unignore them as well:



```js
{
    ignores: [
        '**/*',
        `!src/`,
        `!src/functions/`,
        `!src/functions/health/`,
        '!src/functions/health/get.js'
    ],
},
```

**Reply by canassa** [reply]
@mdjermanovic Thanks for the reply.

Do you have any suggestions on what I should do in this particular case? My problem is that our CI/CD pipeline adds some random directories to our project, causing ESLint to parse them and fail the pipeline.

I want to exclude everything except the .ts files inside the src directory and a few .js files in the root directory.

Manually unignoring each folder seems error-prone, and I am quite sure that other team members will forget to update the ESLint config if they add another directory inside src.

**Reply by mdjermanovic** [reply, marked as answer]
> I want to exclude everything except the .ts files inside the src directory and a few .js files in the root directory.

Maybe something like this would work for your use case:

```js
{
    ignores: [
        '**/*', // ignore everything
        `!src/**/`, // unignore src/ and its subdirectories (but not files)
        `!src/**/*.ts`, // unignore .ts files inside src/ and its subdirectories
        '!prettier.config.js' // unignore root prettier.config.js
        // ... unignore more root .js files
    ],
}
```

**Reply by canassa** [reply]
That works great! Thanks!
